src/libraries/System.Private.CoreLib/src/System/Math.cs

```diff
@@ -1481,5 +1481,251 @@ public static double ScaleB(double x, int n)
+            else
+            {
+                if (val2High != 0)
+                    goto Overflow;
+                // Compute the 'middle' bits of the long multiplication
+                valMid = BigMul(val1High, (uint)j);
+            }
+
+            // See if any bits after bit 32 are set
+            if (Hi32Bits(valMid) != 0)
+                goto Overflow;
+
+            long ret = (long)(BigMul((uint)i, (uint)j) + (valMid << 32));
+
+            // check for overflow
+            if (Hi32Bits((ulong)ret) < (uint)valMid)
+                goto Overflow;
+
+            if (sign >= 0)
+            {
+                // have we spilled into the sign bit?
+                if (ret < 0)
+                    goto Overflow;
+            }
+            else
+            {
+                ret = -ret;
+                // have we spilled into the sign bit?
+                if (ret > 0)
+                    goto Overflow;
+            }
+
+#if DEBUG
+            Debug.Assert(ret == result, $"Multiply overflow got: {ret}, expected: {result}");
+#endif
+            return ret;
+
+        Overflow:
+            ThrowHelper.ThrowOverflowException();
+            return 0;
+        }
+
+        [StackTraceHidden]
+        private static ulong ULongMultiplyOverflow(ulong i, ulong j)
+        {
+            [MethodImpl(MethodImplOptions.AggressiveInlining)]
+            static uint Hi32Bits(ulong a)
+            {
+                return (uint)(a >> 32);
+            }
+
+            // Get the upper 32 bits of the numbers
+            uint val1High = Hi32Bits(i);
+            uint val2High = Hi32Bits(j);
+
+            ulong valMid;
+
+            if (val1High == 0)
+            {
+                if (val2High == 0)
+                    return BigMul((uint)i, (uint)j);
+                // Compute the 'middle' bits of the long multiplication
+                valMid = BigMul(val2High, (uint)i);
+            }
+            else
+            {
+                if (val2High != 0)
+                    goto Overflow;
+                // Compute the 'middle' bits of the long multiplication
+                valMid = BigMul(val1High, (uint)j);
+            }
+
+            // See if any bits after bit 32 are set
+            if (Hi32Bits(valMid) != 0)
+                goto Overflow;
+
+            ulong ret = BigMul((uint)i, (uint)j) + (valMid << 32);
+
+            // check for overflow
+            if (Hi32Bits(ret) < (uint)valMid)
+                goto Overflow;
+
+            Debug.Assert(ret == i * j, $"Multiply overflow got: {ret}, expected: {i * j}");
+            return ret;
+
+        Overflow:
+            ThrowHelper.ThrowOverflowException();
+            return 0;
+        }
+
+        private static double ULongToDouble(ulong val)
+        {
+            double conv = (long)val;
+            if (conv < 0)
+                conv += 4294967296.0 * 4294967296.0;  // add 2^64
+            Debug.Assert(conv >= 0);
+            return conv;
+        }
```

Comment:
Interesting note: I experimented uint64->double cast with integer bit operations. It's about 3x faster than FP ALU on my machine.

Reply:
This doesn't use the FPU on X64 today, only SSE2.

Reply:
I mean 3x faster than this implementation. But I don't expect it to be a consistent result around different CPUs.